### PR TITLE
[FixCode] Add diagnosis/fixit to help users deal with BooleanType's removal

### PR DESCRIPTION
**Pull request by nkcsgexi**
```
Due to migration reasons, types used to conform to BooleanType, which
must contain a member var 'boolValue', now does not convert to Bool. This patch
adds a specific diagnosis/fixit to explicitly invoke 'boolValue' to please the
context types.
```

#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci test macOS

**Comment by akyrtzi** [comment]
LGTM.

**Comment by akyrtzi** [comment]
Do you think we can also handle this case as well:

`_ = true && a`

This gives
`error: cannot convert value of type 'BoolFoo' to expected argument type 'Bool'`

**Comment by nkcsgexi** [comment]
@swift-ci Please smoke test

**Comment by nkcsgexi** [comment]
@akyrtzi Did we see this case in actual codebase?

File: lib/Sema/CSDiag.cpp [hunk 1]
```diff
@@ -2928,6 +2928,22 @@ bool FailureDiagnosis::diagnoseGeneralConversionFailure(Constraint *constraint){
     }
     return true;
   }
+
+  // Due to migration reasons, types used to conform to BooleanType, which
+  // contain a member var 'boolValue', now does not convert to Bool. This block
+  // tries to add a specific diagnosis/fixit to explicitly invoke 'boolValue'.
+  if (toType->isBool()) {
+    auto LookupResult = CS->TC.lookupMember(CS->DC, fromType,
+      DeclName(CS->DC->getASTContext().getIdentifier("boolValue")));
```
Comment (jrose-apple) [inline lib/Sema/CSDiag.cpp]: `CS->TC.Context` gets there a little faster.

File: lib/Sema/CSDiag.cpp [hunk 2]
```diff
@@ -2928,6 +2928,22 @@ bool FailureDiagnosis::diagnoseGeneralConversionFailure(Constraint *constraint){
...
+    if (!LookupResult.empty()) {
+      if (isa<VarDecl>(LookupResult.begin()->Decl)) {
+        diagnose(anchor->getLoc(), diag::types_not_convertible_use_bool_value,
+                 fromType, toType).fixItInsertAfter(anchor->getEndLoc(),
+                                                    ".boolValue");
```
Comment (jrose-apple) [inline lib/Sema/CSDiag.cpp]: We may need parenthesization here (consider `x as! ObjCBool`).

**Comment by jrose-apple** [comment]
This is rdar://problem/27894712, BTW, if not other Radars as well.

File: lib/Sema/CSDiag.cpp (hunk 2 above)
Comment (nkcsgexi) [inline lib/Sema/CSDiag.cpp]: Do we have a util to decide whether to apply parenthesis on an expr?

Comment (jrose-apple) [inline lib/Sema/CSDiag.cpp]: Appending a property is roughly the same as `Expr::canAppendCallParentheses`.

Comment (jrose-apple) [inline lib/Sema/CSDiag.cpp]: We should probably rename that at some point.

**Comment by nkcsgexi** [comment]
@swift-ci Please smoke test

**Comment by nkcsgexi** [comment]
@jrose-apple Comments addressed!

**Comment by nkcsgexi** [comment]
@swift-ci Please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci Please test OS X platform

**Comment by nkcsgexi** [comment]
@swift-ci test macOS

**Pull request merged**